examples/hosts/app-integration/corruption-recovery/src/recoveryAgent.ts

```diff
@@ -0,0 +1,291 @@
+
+        await Promise.all([
+            containerInfo.set(corruptedKey, false),
+            containerInfo.set(orgContainerKey, containerId),
+            containerInfo.set(recoveredContainerKey, ""),
+            containerInfo.set(recoveredByKey, ""),
+            crc.write(markedForRecoveryKey, false),
+            crc.write(recoveredKey, false),
+        ]);
+
+        return c.container.attach();
+    }
+
+    public static async getRecoveryAgent(
+        agentId: string,
+        containerSchema: ContainerSchema,
+    ): Promise<RecoveryAgent> {
+        const clientProps = {
+            connection: connectionConfig,
+        };
+        const client = new AzureClient(clientProps);
+        const c = await client.getContainer(agentId, agentContainerSchema);
+
+        return new RecoveryAgent(
+            agentId,
+            // containerSchema,
+            c.container,
+            c.services.audience,
+        );
+    }
+
+    public get getRecoveryInfo(): RecoveryInfo {
+        // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+        if (!this._orgContainerId) {
+            throw new Error("Cannot retreive status: Missing container ID.");
+        }
+
+        return {
+            originalContainerId: this._orgContainerId,
+            isContainerCorrupted: this._recoveryInfo.get<boolean>(corruptedKey),
+            agentId: this.agentId,
+            recoveryStatus: this._recoveryStatus,
+            isContainerRecovered: this.recovered,
+            recoveryLog: this._recoveryLog,
+            recoveredContainerId: this._recoveryInfo.get<string>(
+                recoveredContainerKey,
+            ),
+            recoveredBy: this._recoveryInfo.get<string>(recoveredByKey),
+        };
+    }
+
+    public get recoveredContainerId(): string | undefined {
+        return this._recoveryInfo.get<string>(recoveredContainerKey);
+    }
+
+    /* Simulated corruption */
+
+    public markCorrupted(): void {
+        this._recoveryInfo.set(corruptedKey, true);
+    }
+
+    /* Recovery */
+
+    public async recoverDoc(cb: ContainerSetupCb): Promise<string | undefined> {
+        this.setRecoveryStatus("KickedOff");
+        this.log("Kicked off recovery.");
+
+        if (!this.markedForRecovery) {
+            this.log("Marked for forecovery.");
+            await this.markForRecovery();
+        }
+
+        if (this.recovered) {
+            this.setRecoveryStatus("Skipped");
+            this.log("Doc already recovered!");
+            return this.recoveredContainerId;
+        }
+
+        this.log("Volonteering for recovery");
+        await this.volunteerForRecovery();
+        if (this.recovered) {
+            this.releaseRecoveryTask();
+            this.setRecoveryStatus("Skipped");
+            this.log("We volonteered, but doc already recovered!");
+            return this.recoveredContainerId;
+        }
+
+        const fluidContainer = await this.recoverContainer();
```

Comment:
By this time this client may have lost connection, and thus lost task. And thus, two clients can be doing that same thing.
Is this expected? Assuming that the rest of the code works more or less normally, and one wins, do we want to cleanup extra file?

Reply:
The intention is one-client wins, but in terms of general correctness that does not seem to be a good path forward. I need to rework this example for first/last client wins, and this should be part of it.